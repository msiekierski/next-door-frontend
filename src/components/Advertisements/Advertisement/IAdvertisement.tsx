interface IAdvertisement {
  idAd: number;
  idAccount: number;
  title: string;
  description: string;
  price: number;
  status?: number;
  updateAdvertisement?: Function;
  removeAdvertisement?: Function;
  housingAssocList?: number[];
}

export default IAdvertisement;
