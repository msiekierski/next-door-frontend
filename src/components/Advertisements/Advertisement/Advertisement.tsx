import React, { FunctionComponent, MouseEvent, useContext, useState } from "react";
import { Button, Card, Form } from "react-bootstrap";
import { AiOutlineEdit } from "react-icons/ai";
import IAdvertisement from "./IAdvertisement";
import { UserContext } from "../../Login/UserContext";
import { deleteAdvertisement, putAdvertisement } from "../../../API/advertisement";

type Props = IAdvertisement;

const Advertisement: FunctionComponent<Props> = ({
  idAd,
  title,
  description,
  status,
  idAccount,
  updateAdvertisement,
  removeAdvertisement,
}) => {
  const user = useContext(UserContext);
  const [isEditing, setIsEditing] = useState(false);
  const [titleEdit, setTitleEdit] = useState(title);
  const [descriptionEdit, setDescriptionEdit] = useState(description);

  const handleCancelEditClick = (e: MouseEvent) => {
    e.preventDefault();
    clearEdit();
    setIsEditing(false);
  };

  const handleSaveEditClick = async (e: MouseEvent) => {
    e.preventDefault();
    await putAdvertisement(idAd, titleEdit, descriptionEdit);
    if (updateAdvertisement) updateAdvertisement(idAd, titleEdit, descriptionEdit);
    clearEdit();
    setIsEditing(false);
  };

  const clearEdit = () => {
    setDescriptionEdit(description);
    setTitleEdit(title);
  };

  const handleDeleteEditClick = async (e: MouseEvent) => {
    e.preventDefault();
    await deleteAdvertisement(idAd);
    if (removeAdvertisement) removeAdvertisement(idAd);
  };

  return (
    <Card className={`mt-3`}>
      <Card.Header className={`bg-success`} />
      <Card.Body>
        <Card.Title className={`d-flex justify-content-between mb-0`}>
          <span>
            {!isEditing ? (
              title
            ) : (
              <Form.Control type="text" value={titleEdit} onChange={(e) => setTitleEdit(e.target.value)} />
            )}
          </span>
          <div className="flex-column">
            <Card.Subtitle className={`text-right mb-3`}>
              {user?.idAccount === idAccount && !isEditing && (
                <span className={`btn `}>
                  <AiOutlineEdit onClick={() => setIsEditing(!isEditing)} />
                </span>
              )}
              {user?.idAccount === idAccount && isEditing && (
                <div className={`d-flex justify-content-end`}>
                  <Button variant="danger" className="mr-2" onClick={handleDeleteEditClick}>
                    Delete
                  </Button>
                  <Button variant="warning" className="mr-2" onClick={handleCancelEditClick}>
                    Cancel
                  </Button>
                  <Button variant="primary" onClick={handleSaveEditClick}>
                    Save
                  </Button>
                </div>
              )}
            </Card.Subtitle>
          </div>
        </Card.Title>
        <Card.Text>
          {!isEditing ? (
            description
          ) : (
            <Form.Control type="text" value={descriptionEdit} onChange={(e) => setDescriptionEdit(e.target.value)} />
          )}
        </Card.Text>
        {status == 0 ? (
          <Button variant={"warning"}>Activate and pay</Button>
        ) : (
          <Card.Subtitle className={`text-success`}>Active</Card.Subtitle>
        )}
      </Card.Body>
    </Card>
  );
};

function lookUpVariant(variant: string) {
  if (variant === "communal") {
    return "primary";
  } else if (variant === "administrative") return "warning";
}

export default Advertisement;
