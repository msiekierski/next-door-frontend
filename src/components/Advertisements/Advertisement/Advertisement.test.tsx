import React from "react";
import renderer from "react-test-renderer";
import Advertisement from "./Advertisement";
import IAdvertisement from "./IAdvertisement";

type Props = IAdvertisement & { status: number };

const props1: Props = {
    idAd: 1,
    idAccount: 1,
    title: "Advertisement",
    description: "ad desc",
    price: 20,
    updateAdvertisement: () => {},
    removeAdvertisement: () => {},
    status: 0
}

const props2: Props = {
  idAd: 1,
  idAccount: 1,
  title: "Advertisement",
  description: "ad desc",
  price: 20,
  updateAdvertisement: () => {},
  removeAdvertisement: () => {},
  status: 1,
};

it("renders correctly unpaid ad", () => {
  const tree = renderer.create(<Advertisement {...props1}/>);
  expect(tree).toMatchSnapshot();
});


it("renders correctly paid ad", () => {
  const tree = renderer.create(<Advertisement {...props2} />);
  expect(tree).toMatchSnapshot();
});
