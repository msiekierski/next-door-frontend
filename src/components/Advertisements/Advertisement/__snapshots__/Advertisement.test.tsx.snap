// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly paid ad 1`] = `
<div
  className="mt-3 card"
>
  <div
    className="bg-success card-header"
  />
  <div
    className="card-body"
  >
    <div
      className="d-flex justify-content-between mb-0 card-title h5"
    >
      <span>
        Advertisement
      </span>
      <div
        className="flex-column"
      >
        <div
          className="text-right mb-3 card-subtitle h6"
        />
      </div>
    </div>
    <p
      className="card-text"
    >
      ad desc
    </p>
    <div
      className="text-success card-subtitle h6"
    >
      Active
    </div>
  </div>
</div>
`;

exports[`renders correctly unpaid ad 1`] = `
<div
  className="mt-3 card"
>
  <div
    className="bg-success card-header"
  />
  <div
    className="card-body"
  >
    <div
      className="d-flex justify-content-between mb-0 card-title h5"
    >
      <span>
        Advertisement
      </span>
      <div
        className="flex-column"
      >
        <div
          className="text-right mb-3 card-subtitle h6"
        />
      </div>
    </div>
    <p
      className="card-text"
    >
      ad desc
    </p>
    <button
      className="btn btn-warning"
      disabled={false}
      type="button"
    >
      Activate and pay
    </button>
  </div>
</div>
`;
